[KED-950] Improve test coverage (branch coverage)

neomatrix369 opened the pull request:
## Description
On back of issue #243 and also taking the work of PR #267 further.

## Development notes
No changes to functionality, writing tests to improve branch coverage and potentially also to take care of the failing Java components due to Java dependencies (to be investigated).

## Checklist

- [X] Read the [contributing](../CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [X] Updated the documentation to reflect the code changes
- [X] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](../RELEASE.md) file
- [X] Added tests to cover my changes

## Notice

- [X] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

neomatrix369 commented:
@ZainPatelQB I have extracted the coverage report of the `kedro` source folder (excluding the `template` subfolder. From the report, it seems there are only 8 (+2 I skipped, but will revert to normal soon) missing branches to be fixed. 

But looking closely, of the 8 missing branches, a handful of them are not yet implemented (methods contain `pass`) and others might not be worth writing tests for.

Can you please quickly navigate through the attached report for me to let me know which ones you would like per se covered and which ones can be added to the omit / do not cover list so we are finished with the PR soonish ;)

[kedro-branch-coverage.zip](https://github.com/quantumblacklabs/kedro/files/4496800/kedro-branch-coverage.zip)

neomatrix369 commented:
The PR is ready to be reviewed, happy to discuss my changes further.

I make some spare time I wrote the missing tests but then found that someone had already written one of them and pushed it to develop today, making my test redundant. 

As it's an F/OSS project this can happen but could we please communicate in future. I had clearly published the list [here](https://github.com/quantumblacklabs/kedro/pull/327#issuecomment-615883140) and didn't hear back about my comment.

mzjp2 commented:
Hey @neomatrix369, super sorry for the radio silence here. It's been a hectic time and this got lost in the backlog. I've moved the ticket to in-review so  we'll assign a few reviewers and get this looked at. Thank you for doing it!

> As it's an F/OSS project this can happen but could we please communicate in future. I had clearly published the list here, and one of the pathways was included in it, but didn't hear back about my comment from anyone.

Fully agree with you here. 🙌

mzjp2 commented on Makefile:
```diff
@@ -19,7 +19,10 @@ lint:
 	pre-commit run -a --hook-stage manual
 
 test:
-	pytest tests
+	@./test-coverage.sh
+
+test-coverage:
+	@./test-coverage.sh
```
are these just duplicates?

neomatrix369 replied:
Well I started off with a new make `rule` and then when working on the CircleCI config I realised that to make sure all CircleCI tasks pick up the coverage shaabang, I decided to change that line too. 

What would you prefer, `test` rule that also does coverage or a new rule that's called `test-coverage`? If I'm right, currently we don't have one.

mzjp2 commented on test-coverage.sh:
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+set -e
+set -u
+set -o pipefail
```
I'm not too big a fan of this file, I haven't read it, but it looks like a lot for something that is just meant to run tests.

neomatrix369 replied:
What do you suggest we do? This helps me when I'm working on any python project, so I don't have to remember all the commands and config files to call. We could move all of this into the make file, but that would bulk it up.

Happy to hear suggestions.

neomatrix369 replied:
Since we will always write tests when we implement something or fix something, having something before us that gives us all the necessary info with just one command (and we don't have to remember all the `pytest` runner args) - thought this could be something to help the developer.

marcelotrevisani replied:
I think you should use tox then :)
It will trigger everything for you
https://tox.readthedocs.io/en/latest/

mzjp2 commented on test_requirements.txt:
```diff
@@ -29,6 +29,7 @@ pyarrow>=0.12.0, <1.0.0
 pylint>=2.4.4, <3.0
 pyspark>=2.2.0, <3.0; python_version < '3.8'
 pytest-cov>=2.5, <3.0
+pytest-html>=2.0.1
```
Can you upper bound this by 3.0?

neomatrix369 replied:
Sure!

neomatrix369 replied:
Done via f8718aa49d17ae7da5386c47340ff045f81ca6ae

neomatrix369 commented:
> Hey @neomatrix369, super sorry for the radio silence here. It's been a hectic time and this got lost in the backlog. I've moved the ticket to in-review so we'll assign a few reviewers and get this looked at. Thank you for doing it!
> 
> > As it's an F/OSS project this can happen but could we please communicate in future. I had clearly published the list here, and one of the pathways was included in it, but didn't hear back about my comment from anyone.
> 
> Fully agree with you here. 🙌

No worries we are all busy-bees, atleast it got to the top of your list ;)

mzjp2 commented:
So, I think the main contribution we want to accept here is the extra tests that improve the test coverage. I think we should leave the `make test` alone and let it do what it did previously, run all tests and print the coverage report to stdout using `pytest-cov` as it did before (no html reports)

The `test-coverage.sh` can go in `tools/test-coverage.sh` and we can have a `make test-coverage` that generates the report by calling the script.

WDYT? :)

neomatrix369 commented:
> So, I think the main contribution we want to accept here is the extra tests that improve the test coverage. I think we should leave the `make test` alone and let it do what it did previously, run all tests and print the coverage report to stdout using `pytest-cov` as it did before (no html reports)
> 
> The `test-coverage.sh` can go in `tools/test-coverage.sh` and we can have a `make test-coverage` that generates the report by calling the script.
> 
> WDYT? :)

Sounds reasonable, let me push some commits and hopefully, it will reflect what you mention above. Thanks for the feedback.

neomatrix369 commented:
@mzjp2 I have started looking into this but now for some reason, things don't build like they used to with the latest commits from `develop`. I'll try to fix them and address your comments.

mzjp2 commented:
Oh, that's weird. I don't think we changed the way we built things since. I know we upgraded pip-tools but that was just to fix things. I'll pop by in a bit and have a look at what happened 🤔

neomatrix369 commented:
> Oh, that's weird. I don't think we changed the way we built things since. I know we upgraded pip-tools but that was just to fix things. I'll pop by in a bit and have a look at what happened 🤔

So it surprises me how the tests broke over the last few days while my last comments were all green, I'll look into this, maybe it part of this enabling of the coverage flags. I'm also trying a new environment to reproduce the issue. I'll keep you posted.

neomatrix369 commented:
> Oh, that's weird. I don't think we changed the way we built things since. I know we upgraded pip-tools but that was just to fix things. I'll pop by in a bit and have a look at what happened 🤔

So to narrow down the problem I created a new conda environment called `kedro`.  Installed everything as mentioned in the docs (the `make` commands). Switched to the `develop` branch, got the latest commits and ran in the project root folder:
```
make test
```

And I get:

```
ImportError while loading conftest '../path/to/conftest/py` ...
```
Similar to https://stackoverflow.com/questions/44067609/getting-error-importmismatcherror-while-running-py-test

We get the same issue with:
```
pytest 
```

The fix is to remove all the **/*.pyc files and `__pycache__` folders across the project. So it's not any project related changes, local environment issues.

mzjp2 commented:
> > Oh, that's weird. I don't think we changed the way we built things since. I know we upgraded pip-tools but that was just to fix things. I'll pop by in a bit and have a look at what happened 🤔
> 
> So to narrow down the problem I created a new conda environment called `kedro`. Installed everything as mentioned in the docs (the `make` commands). Switched to the `develop` branch, got the latest commits and ran in the project root folder:
> 
> ```
> make test
> ```
> 
> And I get:
> 
> ```
> ImportError while loading conftest '../path/to/conftest/py` ...
> ```
> 
> Similar to https://stackoverflow.com/questions/44067609/getting-error-importmismatcherror-while-running-py-test
> 
> We get the same issue with:
> 
> ```
> pytest 
> ```
> 
> The fix is to remove all the **/*.pyc files and `__pycache__` folders across the project. So it's not any project related changes, local environment issues.

Nice job hunting that down. Didn't know that pyc and pycache files could ever cause a problem, good to keep in mind. Glad it's working now! Will get some eyes to review your code tomorrow/Monday hopefully! 🤞

neomatrix369 commented:
> > > Oh, that's weird. I don't think we changed the way we built things since. I know we upgraded pip-tools but that was just to fix things. I'll pop by in a bit and have a look at what happened 🤔
> > 
> > 
> > So to narrow down the problem I created a new conda environment called `kedro`. Installed everything as mentioned in the docs (the `make` commands). Switched to the `develop` branch, got the latest commits and ran in the project root folder:
> > ```
> > make test
> > ```
> > 
> > 
> > And I get:
> > ```
> > ImportError while loading conftest '../path/to/conftest/py` ...
> > ```
> > 
> > 
> > Similar to https://stackoverflow.com/questions/44067609/getting-error-importmismatcherror-while-running-py-test
> > We get the same issue with:
> > ```
> > pytest 
> > ```
> > 
> > 
> > The fix is to remove all the **/*.pyc files and `__pycache__` folders across the project. So it's not any project related changes, local environment issues.
> 
> Nice job hunting that down. Didn't know that pyc and pycache files could ever cause a problem, good to keep in mind. Glad it's working now! Will get some eyes to review your code tomorrow/Monday hopefully! 🤞

We could put this in some troubleshooting docs, so others can look at it when they come across such errors. Easy to happen and time consuming sometimes.

neomatrix369 commented:
@mzjp2  you will notice I have created a new coverage config file `default_coverage_report.toml`, although there is one called `pyproject.toml` already. Not sure where or how it is used, I did a `grep` search and found nothing. We could potentially merge it into one file - although I don't know the impact of this action.

mzjp2 commented on tests/template/test_load_context.py:
```diff
@@ -51,14 +51,42 @@ def test_valid_context(self, fake_repo_path):
+        assert result_first_time.project_version == kedro.__version__
+        assert str(fake_repo_path.resolve() / "src") in sys.path
+
+        # second time
+        try:
+            result_second_time = load_context(str(fake_repo_path))
+            assert result_first_time.project_path == result_second_time.project_path
+            assert (
+                result_first_time.project_version == result_second_time.project_version
+            )
+            assert str(fake_repo_path.resolve() / "src") in sys.path
+        except KedroContextError as kce:
+            print(
+                "FAILED: Should have loaded the context twice successfully, "
+                "and not got this exception:",
+                kce,
+            )
+            assert 0
```
What is this doing? Can't you just load the context twice? If an exception is raised, the test will fail. If you want to provide more information, I'd replace the `print` and `assert 0` with `pytest.fail(your_failure_reason)`

neomatrix369 replied:
Good point I can make it condensed.

neomatrix369 replied:
Applied via 9b9897

DmitriiDeriabinQB commented on default_coverage_report.toml:
```diff
@@ -0,0 +1,11 @@
+fail_under = 100
+show_missing = true
+omit = [
+  "tests/*",
```
We probably shouldn't omit tests coverage

neomatrix369 replied:
Are you sure you want to do that, isn't coverage about implementation code. Adding test coverage could add false positives/negatives to the results and maybe even cause branch coverage to fail if the total scores fell below the threshold.

DmitriiDeriabinQB commented on default_coverage_report.toml:
```diff
@@ -0,0 +1,11 @@
+[tool.coverage.report]
```
Is there any particular reason for creating a new file instead of modifying `pyproject.toml`?

neomatrix369 replied:
As I mentioned above, I didn't know the impact of changing it and happy to merge it as one file.

neomatrix369 replied:
Merged the config into `pyproject.toml` via 8bb4db990ce34aa57d90b0de66214427653278e6

DmitriiDeriabinQB commented on tools/test-coverage.sh:
```diff
@@ -0,0 +1,68 @@
+TARGET_TEST_FOLDERS="$@"
+TARGET_TEST_FOLDERS="${TARGET_TEST_FOLDERS:-"${TESTS_FOLDER}/"}"
+run_test_runner() {
+    echo ""; echo "~~~ Running tests with coverage on branch '$(get_git_branch)'"
```
Super minor, but maybe worth calling it once at the beginning and assigning to a variable? Since we are not doing anything with the branch, it shouldn't change mid way through.

neomatrix369 replied:
Applied via c0d3625adab4a2747355dc2a52e68a2c91a72a5e

DmitriiDeriabinQB commented on tools/test-coverage.sh:
```diff
@@ -0,0 +1,68 @@
+SOURCES_FOLDER=kedro
+TESTS_FOLDER=tests
+TARGET_TEST_FOLDERS="$@"
+TARGET_TEST_FOLDERS="${TARGET_TEST_FOLDERS:-"${TESTS_FOLDER}/"}"
```
I guess, we can remove a trailing slash and the quotes?
```suggestion
TARGET_TEST_FOLDERS="${TARGET_TEST_FOLDERS:-${TESTS_FOLDER}}"
```

DmitriiDeriabinQB commented on tools/test-coverage.sh:
```diff
@@ -0,0 +1,68 @@
+run_test_runner
+print_finish_time
+
+exit ${test_run_exit_code=0}
```
Maybe exit with the same code?

```suggestion
exit ${test_run_exit_code}
```

neomatrix369 replied:
This must have been a typo, meant to return the exit code procured.

DmitriiDeriabinQB commented:
> @mzjp2 you will notice I have created a new coverage config file `default_coverage_report.toml`, although there is one called `pyproject.toml` already. Not sure where or how it is used, I did a `grep` search and found nothing. We could potentially merge it into one file - although I don't know the impact of this action.

`pyproject.toml` is supported as a default coverage config, therefore `make test` didn't need an explicit `--cov-config` option, however it's probably worth adding for search purposes.

mzjp2 commented on tests/template/test_load_context.py:
```diff
@@ -66,24 +66,15 @@ def test_loading_valid_context_twice(self, mocker, fake_repo_path):
-                kce,
+                "and not got this exception:" + str(kce)
             )
             assert 0
```
I don't think you need this?
```suggestion
```

neomatrix369 commented:
> > @mzjp2 you will notice I have created a new coverage config file `default_coverage_report.toml`, although there is one called `pyproject.toml` already. Not sure where or how it is used, I did a `grep` search and found nothing. We could potentially merge it into one file - although I don't know the impact of this action.
> 
> @neomatrix369 `pyproject.toml` is supported as a default coverage config, therefore `make test` didn't need an explicit `--cov-config` option, however it's probably worth adding for search purposes.

Thanks for confirming, I have however merged the config into `pyproject.toml` but we could split it into the concerns: default project settings in `pyproject.toml`, code coverage in `pyproject-code-coverage.toml`  if you like it that way. 

Although would settings in `pyproject.toml`  be overridden if we called/used `pyproject-code-coverage.toml`  instead?

mzjp2 commented on tests/template/test_load_context.py:
```diff
@@ -66,24 +66,15 @@ def test_loading_valid_context_twice(self, mocker, fake_repo_path):
+            load_context(str(fake_repo_path))
         except KedroContextError as kce:
-            print(
+            pytest.fail(
```
👏

neomatrix369 commented:
> Looks good to me. Thanks Dmitrii for taking a more thorough look on the coverage side of things.
> 
> FWIW I agree that we shouldn't omit `tests` from coverage. It's valuable information to know when certain bits of your tests aren't being run and helps you troubleshoot what's happening. Tests should be super simple things without (many, if at all) conditional branches anyway, so I wouldn't be too worried about that

I have just added it back, via fdf97a5166352cbf111182b2bfab2324be09bd42

neomatrix369 commented:
Had to apply https://github.com/quantumblacklabs/kedro/pull/327/commits/d5cc4422934ea3d83fe7c14492c953342a299a10 in order to overcome sort of false positive about coverage of the two tests lines.

DmitriiDeriabinQB commented on kedro/templates/project/{{ cookiecutter.repo_name }}/.coveragerc:
```diff
@@ -1,3 +1,5 @@
+[run]
+branch=True
```
I wouldn't want to turn on branch coverage by default for user projects. I think this has has to be a conscious user choice

neomatrix369 replied:
Don't you think it's a good practise to encourage users to enable code coverage before they start building things? Users can then choose to disable the flag if they don't wish to use it.

In practise it is seen, that if this is switched off, user's won't notice and only very few might enable them.

 I'm thinking a bigger number will keep this enabled, as opposed to switching it on as a conscious choice.

DmitriiDeriabinQB commented on pyproject.toml:
```diff
@@ -13,5 +13,11 @@ default_section = "THIRDPARTY"
 [tool.coverage.report]
```
Shouldn't we have a section
```
[tool.coverage.run]
branch=True
```
in both `pyproject.toml` and `pyproject_no_spark.toml`?

neomatrix369 replied:
So far it didn't seem to be required, I could put it in both of them but maybe redundant.

mzjp2 replied:
We don't have a `.coveragerc` file in our repo. It looks like you've set this in the _template_ `.coveragerc` file which only applies to projects _created_ by Kedro (i.e when you do `kedro new`) and not the actual kedro repo itself.

neomatrix369 replied:
I had a closer look and applied the suggestion, to both the `.toml` files in the root folder. So both of you are correct, my apologies.

I was under the impression it wasn't needed. Until we started explicitly passing the `.toml` from the `make` commands and the shell script.
On a different note, it's best to have one method of config declaration either via the `.coveragerc` or via the `.toml` - any reasons for both types being in the package? Might be legacy if I'm not wrong.

Note: I took over the PR from Cheuk and so I became the new owner of all the commits here, and this didn't come to mind immediately to me.

mzjp2 replied:
As far as I know, we do only have `.toml` -- the `.coveragerc` files is within the template that we use to create peoples Kedro packages. It's all a bit meta, but the Kedro package uses `toml` only and contains a cookiecutter template with a `.coveragerc` file as a "static" file that we use to generate a Kedro project.

neomatrix369 replied:
Thanks for explaining, I didn't know the intent behind them.

As we have seen that it can cause assumptions and discussions.

neomatrix369 commented:
@DmitriiDeriabinQB @mzjp2 Are you guys done with all your comments/feedback before I start responding to them?

neomatrix369 commented on tests/template/test_load_context.py:
```diff
@@ -51,14 +51,32 @@ def test_valid_context(self, fake_repo_path):
+        try:
+            # second time
+            load_context(str(fake_repo_path))
+        except KedroContextError as kce:  # pragma: no cover
```
What is the exception raised when `load_context()` fails, other than `KedroContextError`?

DmitriiDeriabinQB replied:
My point here was that I believe we don't need to catch the error here just to run `pytest.fail()`. I think we should let the error bubble up as pytest will catch it anyways. This will reduce the complexity of test code and number of `# pragma: no cover`s with the same result.

neomatrix369 replied:
What you are suggesting here although might appear to save a few lines of code and in your view reduce complexity, although in my view it opens up the scenarios where the test is not clear to read and understand (introduces assumptions) and also would flood the logs/console with stack traces (making less readable). The paradigm of _setup-action-assert_ and the likes are missing in the proposed format above. Also, by not catching the error(s) in the test we are asking the framework to handle it for us and this can go different ways.

I can expound on these points further if you like me to. Initially, Cheuk also introduced checking for the values loaded (which we should have retained), as I think it is a sensible course of action, to cover for silent failures. By which I mean, what if loading does not result in exceptions but invalid values are loaded and we let it through. While you may say that may not be the test we are working and our code is covering for it. In my view from past experience, doing a bit more in a test is always better than not. Also makes the intent of checking clearer for those who come after us. There will always be tests that marginally overall in intent with another bunch of them.

I can give you examples of outputs and other scenarios if you like me to but it would only make this PR longer than it should have been.

I'm happy to cover for the additional tests that have opened up with the branch flag being enabled - hopefully, we can move forward with this one. I propose the below:

```python
    def test_loading_valid_context_twice(self, mocker, fake_repo_path):
        """Test getting project context twice."""
        mocker.patch("logging.config.dictConfig")

        # first time
        try:
            # first time
            load_context(str(fake_repo_path))
            ---> (assert if the loaded context contains valid values, Cheuk's original work) <----
        except KedroContextError as kce:  # pragma: no cover
            pytest.fail(  # pragma: no cover
                "FAILED: Should have loaded the context the first-time successfully (1 out of 2 attempts), "
                "and not got this exception: " + str(kce)
            )
        load_context(str(fake_repo_path))

        try:
            # second time
            load_context(str(fake_repo_path))
            ---> (assert if the loaded context contains valid values, Cheuk's original work) <----
        except KedroContextError as kce:  # pragma: no cover
            pytest.fail(  # pragma: no cover
                "FAILED: Should have loaded the context twice successfully (2 out of 2 attempts), "
                "and not got this exception: " + str(kce)
            )
```
_(with of course applying refactorings of semantically repeated code above)_

I like `kedro`, and everyone else as well. I would not like to introduce gaps in it when I can see it in sight.

DmitriiDeriabinQB commented:
> @DmitriiDeriabinQB @mzjp2 Are you guys done with all your comments/feedback before I start responding to them?

@neomatrix369 Sorry for fragmented reviews, took me several passes to go through everything. I think I've raised all the points that I had.

Regarding enabling branch coverage for the project template - I've asked for another pair of eyes on this, however it's a minor point anyways, so shouldn't block you.

DmitriiDeriabinQB commented on tools/test-coverage.sh:
```diff
@@ -0,0 +1,66 @@
+run_test_runner() {
+    echo ""; echo "~~~ Running tests with coverage on branch '${CURRENT_GIT_BRANCH}'"
+    set -x
+    pytest --cov-config default_coverage_report.toml     \
```
Shouldn't it point to `pyproject.toml`?

neomatrix369 replied:
Yes I will have this changed, it was on my mind, and slipped away.

neomatrix369 commented on CONTRIBUTING.md:
````diff
@@ -70,13 +70,34 @@ If you picked up a code or documentation related issue, before pushing the branc
+.test-run-reports
+```
+
+Note: `make test` will run the tests and generate a terminal-only text coverage report both on the local machine and on CircleCI.
````
I'll change this, good point.

neomatrix369 replied:
I have amended this, please take a look.

neomatrix369 commented on Makefile:
```diff
@@ -19,7 +19,10 @@ lint:
 
 test:
-	pytest tests
+	pytest tests --cov=kedro
```
Will remove asap.

921kiyo commented:
Hi, thank you for your work and contribution! As the codebase has changed quite a bit since May and so is the scope of this task, I'm closing this PR for now, but feel free to reopen a PR if anyone wants to pick this up :)
